### DISCUSSION TITLE
Large File Upload Takes Extremely Long then Fails

### DISCUSSION THREAD

**Discussion by Dan6erbond** [pocketbase/pocketbase#2417] [Q&A] [answered]
I have a small app that relies on PB file uploading for its main functionality. I followed the Js example using `FormData` to upload the file, which works well for smaller size images (<4MB).

Anything larger than that, especially raw photos from my phone's camera or videos take extremely long, and seem to fail after a while with this error:

```json
{"code":400,"message":"Failed to load the submitted data due to invalid formatting.","data":{}}
```

The request timing is also pretty bad. I'm using MinIO as an S3 backend if it makes a difference:

![image](https://user-images.githubusercontent.com/32235074/236142457-e02ed0bc-c0a9-46a0-a3e0-e4b4ab3c3c84.png)

My payload looks like this:

```formdata
------WebKitFormBoundaryOHvddAEgBOjIoGLr
Content-Disposition: form-data; name="file"; filename="2022 11 04 - Oberalppass Arrival.mp4"
Content-Type: video/mp4


------WebKitFormBoundaryOHvddAEgBOjIoGLr
Content-Disposition: form-data; name="name"

2022 11 04 - Oberalppass Arrival.mp4
------WebKitFormBoundaryOHvddAEgBOjIoGLr
Content-Disposition: form-data; name="type"

video/mp4
------WebKitFormBoundaryOHvddAEgBOjIoGLr
Content-Disposition: form-data; name="author"

2s4h5f7u2d10fxc
------WebKitFormBoundaryOHvddAEgBOjIoGLr
Content-Disposition: form-data; name="description"


------WebKitFormBoundaryOHvddAEgBOjIoGLr--
```

This also happens in the PocketBase admin UI.

Why is this issue occurring? How can I improve upload performance/show a loading indicator to improve the UX?

**Comment by Dan6erbond** [comment]
So I did some investigating, and tested with a local instance of PocketBase available at 127.0.0.1:8080 and compared the upload speeds against the version hosted on my server, with a drastic difference in upload times. The local version was almost instant. 

Beyond that, it looks like I had the file size limit set too low. So I guess that explains that issue. I'll have to find out why the speeds are so slow, since my internet speed should be able to keep up with a 30MB video file. 😅

**Comment by ganigeorgiev** [comment, marked as answer]
In general the upload time will depend on client network speed and whether the external storage (in your case MinIO) is in the same region as the PocketBase server.

> Why is this issue occurring? 

I'm not sure. We have a default read timeout duration of 5mins (_I'll consider making it configurable in the future_). Another thing could be also a timeout on reverse proxy level (eg. nginx configuration).

> How can I improve upload performance/show a loading indicator to improve the UX?

~A 4MB file is not really that big so I'm not sure why is so slow in your case.~ _On second read, your issue seems to be for files larger than 4MB._

The easiest way to speed up the upload is to use a local filesystem storage instead of S3 to eliminate the network latency between PocketBase and MinIO.

**Comment by Dan6erbond** [comment]
@ganigeorgiev thanks for the tips! You're definitely right that MinIO is adding some latency for the file upload. Unfortunately I have PocketBase deployed in a container, so was trying to avoid using the local filesystem, but I'll see how bad the performance really is.

MinIO and PocketBase are running on the same network so to speak, as they're on the same cluster. I tried an internal URL to no avail. The speeds are roughly the same, which is leading me to believe that the main performance hit is just bandwidth either on the client side or the server side (though I really doubt the server internet speeds aren't sufficient for what I'm doing).

Also disabling S3 entirely didn't really change much. Locally it works well, after I increase the file size limit to 10GB which I'm comfortable with, the uploads are extremely fast. But that makes sense given there's no traffic going over the internet.

It's a bit frustrating to see that this is most likely an issue with infrastructure and not code haha, but I'm glad it's not an issue with PocketBase so I'll see what I can do to improve performance!

**Comment by Maxson52** [comment]
I'm having this issue on my local dev server with files uploading to the local file system storage, not to S3.

```js
const data = Object.fromEntries(await request.formData()) as {
  image: File;
};
try {
  await locals.pb.collection("logos").create(data);
} catch (err) {
  console.log(err);
}
```
This is inside a SvelteKit action which runs until it times out and the error mentioned in the OP is shown in the logs.

**Reply by ganigeorgiev** [reply]
I don't see how your issue is related to the above discussion (or to PocketBase since I'm not able to reproduce it when testing locally).

The easiest way to rule out PocketBase issue is to try send direct `multipart/form-data` request with Postman, Insomnia, curl, etc. to the PocketBase server without SvelteKit.

Most likely your issue is misconfiguration of your SvelteKit server, docker/reverse proxy (if use any) or wrong data format. See also the body size limit (default of 512kb) that you may have to adjust - https://kit.svelte.dev/docs/adapter-node#environment-variables-bodysizelimit.

If you still think that it is PocketBase related, then please open a new Q&A discussion with more details about your setup, collections export (Admin UI > Settings > Export collections) and more detailed code sample showing how the SDK was initialized and what `data` is in the above.
